src/sp/it/pl/plugin/LibraryWatcher.kt

```diff
@@ -0,0 +1,152 @@
+                sourceDirs.onItemRemoved { handleLocationRemoved(it) }
+        )
+    }
+
+    @IsConfig(name = "Update on start", info = "Update library when this plugin starts")
+    val updateOnStart by cv(false)
+
+    @IsConfig(name = "Monitoring supported", info = "On some system, this file monitoring may be unsupported", editable = NONE)
+    val dirMonitoringSupported by c(Os.WINDOWS.isCurrent)
+
+    @IsConfig(name = "Monitor files", info = "Monitors files recursively, notify of changes and update library automatically")
+    val dirMonitoringEnabled by cv(false).readOnlyUnless(dirMonitoringSupported)
+
+    private val dirMonitors = HashMap<File, FileMonitor>()
+    private val dirMonitoring = when {
+        dirMonitoringSupported -> Subscribed {
+            Subscription.multi(
+                    dirMonitoringEnabled sync { sourceDirsChangeHandler.subscribe(it) },
+                    Subscription { sourceDirsChangeHandler.subscribe(false) }
+            )
+        }
+        else -> Subscribed { Subscription { } }
+    }
+    private val toBeAdded = HashSet<File>()
+    private val toBeRemoved = HashSet<File>()
+    private val update = EventReducer.toLast<Void>(2000.0) { it -> updateLibraryFromEvents() }
+
+    override fun onStart() {
+        dirMonitoring.subscribe(true)
+        if (updateOnStart.value) updateLibrary()
+    }
+
+    override fun onStop() {
+        dirMonitoring.subscribe(false)
+        dirMonitors.values.forEach { it.stop() }
+        dirMonitors.clear()
+        updateLibraryFromEvents()
+    }
+
+    private fun handleLocationRemoved(dir: File) {
+        if (!dirMonitoringSupported && !dirMonitoringEnabled.value) return
+
+        val wasDuplicate = dir in sourceDirs
+        if (!wasDuplicate) {
+            dirMonitors.remove(dir)?.stop()
+            sourceDirs.forEach { handleLocationAdded(it) }  // starts monitoring previously shadowed directories
+        }
+    }
+
+    private fun handleLocationAdded(dir: File) {
+        if (!dirMonitoringSupported && !dirMonitoringEnabled.value) return
+
+        val isDuplicate = dir in dirMonitors.keys
+        val isShadowed = dirMonitors.keys.any { monitoredDir -> dir isChildOf monitoredDir }
+        val needsMonitoring = !isDuplicate && !isShadowed
+        if (needsMonitoring) {
+            dirMonitors[dir] = monitorDirectory(dir, true) { type, file ->
+                when (type) {
+                    ENTRY_CREATE -> runFX {
+                        toBeAdded += file
+                        update.push(null)
+                    }
+                    ENTRY_DELETE -> runFX {
+                        toBeRemoved += file
+                        update.push(null)
+                    }
+                }
+            }
+        }
+    }
+
+    private fun updateLibraryFromEvents() {
+        val toAdd = toBeAdded.materialize()
+        val toRem = toBeRemoved.materialize()
+        toBeAdded.clear()
```

Comment:
The materializing and clearing of the lists needs to be one atomic operation, unless all modifications to the sets and `updateLibraryFromEvents` always run on the FX thread

Reply:
Will have to check, but I think you are wrong. Materialize is like making a snapshot copy of the list to freeze it in time and provide a static unchanging access free of concurrent modification errors.

Reply:
All FX thread, no problem.

Reply:
Ok. The issue I was talking about here was that you have to assume the worst in a world of multi-threading, so there could have been items added to the List between `materialize` and `clear`, which would then have simply fallen out. But if it all runs on FX no problem. Maybe add a `failIfNonFX` at the beginning of the function to make it clear.

Reply:
No problem. Good to double check.

The check is not necessary, this is trivial case. Everything in the plugin is FX thread except few isolated lines at the bottom, which have clear boundaries (read do not execute any plugin's methods). The thread checks are fine if the boundaries are not clear. Not the case.